src/java/org/apache/cassandra/transport/CQLMessageHandler.java

```diff
@@ -466,13 +465,34 @@ protected boolean processFirstFrameOfLargeMessage(IntactFrame frame, Limit endpo
             
             LargeMessage largeMessage = new LargeMessage(header);
 
-            if (DatabaseDescriptor.getNativeTransportRateLimitingEnabled())
+            if (!acquireCapacity(header, endpointReserve, globalReserve))
```

Comment:
Is it definitely fine to acquire capacity here, if we may throw on overload for rate limiter afterwards? Would be nice to annotate with a comment justifying the safety.

Reply:
Absolutely agree with adding a few comments. It looks like we release in the exception handling path, but I've also added a few assertions and a new test case in `RateLimitingTest` along w/ this commit to put some guardrails around that.